Fix for deserializing imported ordered dictionary

davidBar-On opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

<!-- Summarize your PR between here and the checklist. -->
Suggested fix for issue #2861 - exported ordered dictionary is now imported via `Import-Clixml` as `OrderedDictionary` and not as `Hashtable`.  This is done by checking if the imported dictionary type names include type of `OrderedDictionary`.

## PR Context

<!-- Provide a little reasoning as to why this Pull Request helps and why you have opened it. -->
Fix #2861.

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed: <!-- Number/link of that issue here -->

davidBar-On committed:
Fix issue #2861 - deserialize imported ordered dictionary as ordered dictionary

iSazonov commented on src/System.Management.Automation/engine/serialization.cs:
```diff
@@ -3992,14 +3993,27 @@ private object ReadListContainer(ContainerType ct)
             Hashtable table = new Hashtable(StringComparer.CurrentCultureIgnoreCase);
+            OrderedDictionary tableOrdered = new OrderedDictionary(StringComparer.CurrentCultureIgnoreCase);
+
+            // Find whether original directory was odered or not
+            bool isOrdered = false;
+            if (InternalTypeNames.Count > 0)
+            {
+                string originalDictTypeName = Deserializer.MaskDeserializationPrefix(InternalTypeNames[0]);
+                if (originalDictTypeName == tableOrdered.GetType().FullName)
+                {
+                    isOrdered = true;
+                }
+            }
```
There is a typo in comment - "odered".

Also we could pack the code in one line initialization:
```suggestion
            // Find whether original directory was ordered or not
            bool isOrdered = InternalTypeNames.Count > 0 && (Deserializer.MaskDeserializationPrefix(InternalTypeNames[0]) == tableOrdered.GetType().FullName));
```

davidBar-On replied:
Changed.

iSazonov commented on src/System.Management.Automation/engine/serialization.cs:
```diff
@@ -3992,14 +3993,27 @@ private object ReadListContainer(ContainerType ct)
             // a case insensitive string comparer.  If we discover a key collision,
             // we'll revert back to the default comparer.
             Hashtable table = new Hashtable(StringComparer.CurrentCultureIgnoreCase);
+            OrderedDictionary tableOrdered = new OrderedDictionary(StringComparer.CurrentCultureIgnoreCase);
```
Why do we name the variable as table if it is a dictionary? Maybe `orderedDictionary`.

davidBar-On replied:
Name changed.

iSazonov commented on src/System.Management.Automation/engine/serialization.cs:
```diff
@@ -4039,38 +4053,78 @@ private object ReadDictionary(ContainerType ct)
 
-                    // On the first collision, copy the hash table to one that uses the default comparer.
-                    if (table.ContainsKey(key) && (keyClashFoundIteration == 0))
+                    if (!isOrdered)
```
Why don't swap and start with `if (isOrdered)`?

davidBar-On replied:
Swapped.

iSazonov commented on src/System.Management.Automation/engine/serialization.cs:
```diff
@@ -4039,38 +4053,78 @@ private object ReadDictionary(ContainerType ct)
-                        Hashtable newHashTable = new Hashtable();
-                        foreach (DictionaryEntry entry in table)
+                        // On the first collision, copy the hash table to one that uses the default comparer.
+                        if (table.ContainsKey(key) && (keyClashFoundIteration == 0))
```
```suggestion
                        if (keyClashFoundIteration == 0 && table.ContainsKey(key))
```

Below too.

davidBar-On replied:
This was a copy from the original code - changed in all 4 places (not/ordered x iteration=0/1).

iSazonov commented on src/System.Management.Automation/engine/serialization.cs:
```diff
@@ -4039,38 +4053,78 @@ private object ReadDictionary(ContainerType ct)
                         {
-                            newHashTable.Add(entry.Key, entry.Value);
+                            keyClashFoundIteration++;
+                            Hashtable newHashTable = new Hashtable();
```
```suggestion
                            Hashtable newHashTable = new Hashtable(table.Count);
```

Below too.

davidBar-On replied:
Changed in both cases (again, was in the original code).

iSazonov commented on src/System.Management.Automation/engine/serialization.cs:
```diff
@@ -4039,38 +4053,78 @@ private object ReadDictionary(ContainerType ct)
                         {
-                            newHashTable.Add(entry.Key, entry.Value);
+                            keyClashFoundIteration++;
+                            OrderedDictionary newOrderedTable = new OrderedDictionary();
```
```suggestion
                            OrderedDictionary newOrderedTable = new OrderedDictionary(tableOrdered.Count);
```

davidBar-On replied:
Changed in both cases.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:
```diff
@@ -52,6 +52,31 @@ Describe "XmlCommand DRT basic functionality Tests" -Tags "CI" {
+		$dict['Moe'] = 'Wise guy!'
+		$dict | Export-Clixml $testfile
+		$results = Import-Clixml $testfile
+		$results.GetType().Name | Should -Be "Hashtable"
```
Please use `BeExactly` for strings:
```suggestion
		$results.GetType().Name | Should -BeExactly "Hashtable"
```

davidBar-On replied:
Changed.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:
```diff
@@ -52,6 +52,31 @@ Describe "XmlCommand DRT basic functionality Tests" -Tags "CI" {
+		$dict['Moe'] = 'Wise guy!'
+		$dict | Export-Clixml $testfile
+		$results = Import-Clixml $testfile
+		$results.GetType().Name | Should -Be "OrderedDictionary"
```
```suggestion
		$results.GetType().Name | Should -BeExactly "OrderedDictionary"
```

davidBar-On replied:
Changed.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:
```diff
@@ -59,8 +59,8 @@ Describe "XmlCommand DRT basic functionality Tests" -Tags "CI" {
-		$results.GetType().Name | Should -Be "Hashtable"
-		$results.Count | Should -Be 3
+		$results.GetType().Name | Should -BeExactly  "Hashtable"
+		$results.Count | Should -BeExactly 3
```
It is not string.
```suggestion
		$results.Count | Should -Be 3
```

davidBar-On replied:
Done

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1:
```diff
@@ -70,11 +70,11 @@ Describe "XmlCommand DRT basic functionality Tests" -Tags "CI" {
-		$results[1] | Should -Be $dict[1]
-		$results[2] | Should -Be $dict[2]
+		$results.GetType().Name | Should -BeExactly  "OrderedDictionary"
+		$results.Count | Should -BeExactly  3
```
```suggestion
		$results.Count | Should -Be  3
```

davidBar-On replied:
Done

davidBar-On committed:
Changes per code review comments (2)

daxian-dbw commented:
@PaulHigin can you please take a look at this PR and see the change to `serailization.cs` is the right fix?

PaulHigin reviewed (changes requested):
These changes look right to me, except that instead of having if/else blocks, I feel it would be cleaner if you created a helper class that encapsulates the dictionary types (Hashtable or OrderedDictionary) and handles adding key/value pairs and re-creating with case sensitive comparer as needed.  Something like:

```csharp
private class PSDictionary
{
    private Hashtable table;
    private OrderedDictionary orderedDictionary;

    public PSDictionary(bool isOrdered) { ... }

    public Add(object key, object value) { ... }

    public object DictionaryObject { get { return _isOrdered ? orderedDictionary : table; } }
}
```

Also, I think it is fine to compare the type name check to a hard coded constant string: `System.Collections.Specialized.OrderedDictionary`, instead of getting the full type name from an instance of the object.

davidBar-On committed:
Merge branch 'master' into issue-2861-deserialize-imported-ordered-dictionary

davidBar-On commented:
> ... instead of having if/else blocks, I feel it would be cleaner if you created a helper class .....

Done

> Also, I think it is fine to compare the type name check to a hard coded constant string: `System.Collections.Specialized.OrderedDictionary`, instead of getting the full type name from an instance of the object.

Did not make the this change, as it seems to me that it safer not to assume the type name.  Note that with the other changes `FullName` is now for the class `typeOf` and not for a variable.

davidBar-On committed:
Enhancements per review comments including adding tests for keys clash cases

davidBar-On commented on src/System.Management.Automation/engine/serialization.cs:
```diff
@@ -3986,19 +3987,104 @@ private object ReadListContainer(ContainerType ct)
+            public void Add(object key, object value) {
+                // On the first collision, copy the hash table to one that uses the default comparer.
+                _containsKey = _isOrdered ? orderedDictionary.Contains(key) : table.ContainsKey(key);
+                if (_keyClashFoundIteration == 0 && _containsKey)
```
Added test cases.  See the general discussion comments.

davidBar-On commented:
> Can you add a test for a key clash? Both kinds

Added test cases for the keys clash.  However, for most of the cases I created an "exported" XML manually, as I did not find a way to create hashtable (ordered or not) that includes duplicate keys, or an ordered hashtable that allows case-sensitive keys.  The only added test without using XML is for non-ordered case-sensitive hashtable using `New-Object hashtable` instead of `@{}`.

In the added tests I didn't find a way to check which "Comparer" is set for the imported table.  Therefore the tests only check the number of elements in the imported table.

Also note that I rewrote the change using the `IDictionary` interface to simplify the implementation.

SteveL-MSFT commented on src/System.Management.Automation/engine/serialization.cs:
```diff
@@ -3986,19 +3987,81 @@ private object ReadListContainer(ContainerType ct)
+            public PSDictionary(bool isOrdered) {
+                _isOrdered = isOrdered;
+
+                // By default use a non cases-sensitive comparer
```
```suggestion
                // By default use a non case-sensitive comparer
```

SteveL-MSFT committed:
Update src/System.Management.Automation/engine/serialization.cs

PaulHigin reviewed (approved):
LGTM

iSazonov commented:
@davidBar-On Thanks for your contribution!
